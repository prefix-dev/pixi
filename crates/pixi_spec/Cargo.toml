[package]
authors.workspace = true
description = "Provides a Rust representation of a Pixi spec"
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "pixi_spec"
readme.workspace = true
repository.workspace = true
version = "0.1.0"

[dependencies]
dirs = { workspace = true }
file_url = { workspace = true }
rattler_conda_types = { workspace = true }
rattler_digest = { workspace = true, features = ["serde"] }
serde = { workspace = true }
serde-untagged = { workspace = true }
serde_with = { workspace = true }
thiserror = { workspace = true }
toml_edit = { workspace = true, features = ["serde"] }
typed-path = { workspace = true }
url = { workspace = true }

[dev-dependencies]
insta = { workspace = true, features = ["yaml", "filters"] }
serde_json = { workspace = true, features = ["preserve_order"] }
serde_yaml = { workspace = true }
