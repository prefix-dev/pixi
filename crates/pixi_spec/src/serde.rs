use std::{fmt::Display, path::PathBuf};

use rattler_conda_types::{
    BuildNumberSpec, ChannelConfig, NamedChannelOrUrl, NamelessMatchSpec,
    ParseStrictness::{Lenient, Strict},
    StringMatcher, VersionSpec,
};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
use serde_with::serde_as;
use url::Url;

use crate::{DetailedSpec, GitReference, GitSpec, PathSpec, PixiSpec, UrlSpec};

#[serde_as]
#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
#[serde(rename_all = "kebab-case")]
struct RawSpec {
    /// The version spec of the package (e.g. `1.2.3`, `>=1.2.3`, `1.2.*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub version: Option<VersionSpec>,

    /// The URL of the package
    pub url: Option<Url>,

    /// The git url of the package
    pub git: Option<Url>,

    /// The path to the package
    pub path: Option<String>,

    /// The git revision of the package
    pub branch: Option<String>,

    /// The git revision of the package
    pub rev: Option<String>,

    /// The git revision of the package
    pub tag: Option<String>,

    /// The build string of the package (e.g. `py37_0`, `py37h6de7cb9_0`, `py*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build: Option<StringMatcher>,

    /// The build number of the package
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build_number: Option<BuildNumberSpec>,

    /// Match the specific filename of the package
    pub file_name: Option<String>,

    /// The channel of the package
    pub channel: Option<NamedChannelOrUrl>,

    /// The subdir of the channel
    pub subdir: Option<String>,

    /// The md5 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

/// Returns a more helpful message when a version spec is used incorrectly.
fn version_spec_error<T: Into<String>>(input: T) -> Option<impl Display> {
    let input = input.into();
    if input.starts_with('/')
        || input.starts_with('.')
        || input.starts_with('\\')
        || input.starts_with("~/")
    {
        return Some(format!("it seems you're trying to add a path dependency, please specify as a table with a `path` key: '{{ path = \"{input}\" }}'"));
    }

    if input.contains("git") {
        return Some(format!("it seems you're trying to add a git dependency, please specify as a table with a `git` key: '{{ git = \"{input}\" }}'"));
    }

    if input.contains("://") {
        return Some(format!("it seems you're trying to add a url dependency, please specify as a table with a `url` key: '{{ url = \"{input}\" }}'"));
    }

    if let Ok(match_spec) = NamelessMatchSpec::from_str(&input, Lenient) {
        let spec = PixiSpec::from_nameless_matchspec(
            match_spec,
            &ChannelConfig::default_with_root_dir(PathBuf::default()),
        );
        return Some(format!(
            "expected a version specifier but looks like a matchspec, did you mean {}?",
            spec.to_toml_value()
        ));
    };

    if input.contains("subdir") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `subdir` key: '{ version = \"<VERSION_SPEC>\", subdir = \"<SUBDIR>\" }'".to_string());
    }

    if input.contains("channel") || input.contains("::") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `channel` key: '{ version = \"<VERSION_SPEC>\", channel = \"<CHANNEL>\" }'".to_string());
    }

    if input.contains("md5") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `md5` key: '{ version = \"<VERSION_SPEC>\", md5 = \"<MD5>\" }'".to_string());
    }

    if input.contains("sha256") {
        return Some("it seems you're trying to add a detailed dependency, please specify as a table with a `sha256` key: '{ version = \"<VERSION_SPEC>\", sha256 = \"<SHA256>\" }'".to_string());
    }

    None
}

impl<'de> Deserialize<'de> for PixiSpec {
    fn deserialize<D>(deserializer: D) -> Result<PixiSpec, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .expecting(
                "a version string like \">=0.9.8\" or a detailed dependency like { version = \">=0.9.8\" }",
            )
            .string(|str| {
                VersionSpec::from_str(str, Strict)
                    .map_err(|_err|{
                        if let Some(msg) = version_spec_error(str) {
                            serde_untagged::de::Error::custom(msg)
                        } else {
                            serde_untagged::de::Error::custom("invalid version specifier")
                        }
                                            })
                    .map(PixiSpec::Version)
            })
            .map(|map| {
                let raw_spec: RawSpec = map.deserialize()?;

                if raw_spec.git.is_none()
                    && (raw_spec.branch.is_some()
                        || raw_spec.rev.is_some()
                        || raw_spec.tag.is_some())
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`branch`, `rev`, and `tag` are only valid when `git` is specified",
                    ));
                }

                let is_git = raw_spec.git.is_some();
                let is_path = raw_spec.path.is_some();
                let is_url = raw_spec.url.is_some();


                let git_key = is_git.then_some("`git`");
                let path_key = is_path.then_some("`path`");
                let url_key = is_url.then_some("`url`");
                let non_detailed_keys = [git_key, path_key, url_key]
                    .into_iter()
                    .flatten()
                    .collect::<Vec<_>>()
                    .join(", ");

                if !non_detailed_keys.is_empty() && raw_spec.version.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`version` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.build.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.build_number.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.file_name.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.channel.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                if !non_detailed_keys.is_empty() && raw_spec.subdir.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`build` cannot be used with {non_detailed_keys}"),
                    ));
                }

                let non_url_keys = [git_key, path_key]
                    .into_iter()
                    .flatten()
                    .collect::<Vec<_>>()
                    .join(", ");
                if !non_url_keys.is_empty() && raw_spec.sha256.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`sha256` cannot be used with {non_url_keys}"),
                    ));
                }
                if !non_url_keys.is_empty() && raw_spec.md5.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        format!("`md5` cannot be used with {non_url_keys}"),
                    ));
                }

                let spec = match (raw_spec.url, raw_spec.path, raw_spec.git) {
                    (Some(url), None, None) => PixiSpec::Url(UrlSpec {
                        url,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, Some(path), None) => PixiSpec::Path(PathSpec { path: path.into() }),
                    (None, None, Some(git)) => {
                        let rev = match (raw_spec.branch, raw_spec.rev, raw_spec.tag) {
                            (Some(branch), None, None) => Some(GitReference::Branch(branch)),
                            (None, Some(rev), None) => Some(GitReference::Rev(rev)),
                            (None, None, Some(tag)) => Some(GitReference::Tag(tag)),
                            (None, None, None) => None,
                            _ => {
                                return Err(serde_untagged::de::Error::custom(
                                    "only one of `branch`, `rev`, or `tag` can be specified",
                                ));
                            }
                        };
                        PixiSpec::Git(GitSpec { git, rev })
                    },
                    (None, None, None) => {
                        let is_detailed =
                            raw_spec.version.is_some() ||
                                raw_spec.build.is_some() ||
                                raw_spec.build_number.is_some() ||
                                raw_spec.file_name.is_some() ||
                                raw_spec.channel.is_some() ||
                                raw_spec.subdir.is_some() ||
                                raw_spec.md5.is_some() ||
                                raw_spec.sha256.is_some();
                        if !is_detailed {
                            return Err(serde_untagged::de::Error::custom(
                                "one of `version`, `build`, `build-number`, `file-name`, `channel`, `subdir`, `md5`, `sha256`, `git`, `url`, or `path` must be specified",
                            ))
                        }

                        PixiSpec::DetailedVersion(DetailedSpec {
                            version: raw_spec.version,
                            build: raw_spec.build,
                            build_number: raw_spec.build_number,
                            file_name: raw_spec.file_name,
                            channel: raw_spec.channel,
                            subdir: raw_spec.subdir,
                            md5: raw_spec.md5,
                            sha256: raw_spec.sha256,
                        })
                    }
                    (_, _, _) => {
                        return Err(serde_untagged::de::Error::custom(
                            "only one of `url`, `path`, or `git` can be specified",
                        ))
                    }
                };

                Ok(spec)
            })
            .deserialize(deserializer)
    }
}

impl<'de> Deserialize<'de> for PathSpec {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Raw {
            path: String,
        }

        Raw::deserialize(deserializer).map(|raw| PathSpec {
            path: raw.path.into(),
        })
    }
}

impl Serialize for PathSpec {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw {
            path: String,
        }

        Raw {
            path: self.path.to_string(),
        }
        .serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use serde::Serialize;
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn test_round_trip() {
        let examples = [
            json! { "1.2.3" },
            json!({ "version": "1.2.3" }),
            json!({ "version": "1.2.3", "build-number": ">=3" }),
            json! { "*" },
            json!({ "path": "foobar" }),
            json!({ "path": "~/.cache" }),
            json!({ "subdir": "linux-64" }),
            json!({ "channel": "conda-forge", "subdir": "linux-64" }),
            json!({ "channel": "conda-forge", "subdir": "linux-64" }),
            json!({ "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "version": "1.2.3", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main" }),
            // Errors:
            json!({ "ver": "1.2.3" }),
            json!({ "path": "foobar", "version": "1.2.3" }),
            json!({ "version": "//" }),
            json!({ "path": "foobar", "version": "//" }),
            json!({ "path": "foobar", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main", "tag": "v1" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json! { "/path/style"},
            json! { "./path/style"},
            json! { "\\path\\style"},
            json! { "~/path/style"},
            json! { "https://example.com"},
            json! { "https://github.com/conda-forge/21cmfast-feedstock"},
            json! { "1.2.3[subdir=linux-64]"},
            json! { "1.2.3[channel=conda-forge]"},
            json! { "conda-forge::1.2.3"},
            json! { "1.2.3[md5=315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3]"},
            json! { "1.2.3[sha256=315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3]"},
            json! { "*cpu*"},
            json! { "*=*openblas"},
        ];

        #[derive(Serialize)]
        struct Snapshot {
            input: Value,
            result: Value,
        }

        let mut snapshot = Vec::new();
        for input in examples {
            let spec: Result<PixiSpec, _> = serde_json::from_value(input.clone());
            let result = match spec {
                Ok(spec) => serde_json::to_value(&spec).unwrap(),
                Err(e) => {
                    json!({
                        "error": format!("ERROR: {e}")
                    })
                }
            };

            snapshot.push(Snapshot { input, result });
        }

        insta::assert_yaml_snapshot!(snapshot);
    }
}
