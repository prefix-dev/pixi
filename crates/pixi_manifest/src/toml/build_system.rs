use itertools::Either;
use pixi_spec::TomlSpec;
use rattler_conda_types::NamedChannelOrUrl;
use serde::Deserialize;

use crate::{
    build_system::BuildBackend,
    utils::{package_map::UniquePackageMap, PixiSpanned},
    BuildSystem, TomlError,
};

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct TomlBuildSystem {
    pub build_backend: PixiSpanned<TomlBuildBackend>,

    #[serde(default)]
    pub channels: Option<PixiSpanned<Vec<NamedChannelOrUrl>>>,

    #[serde(default)]
    pub additional_dependencies: UniquePackageMap,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct TomlBuildBackend {
    pub name: PixiSpanned<rattler_conda_types::PackageName>,

    #[serde(flatten)]
    pub spec: TomlSpec,
}

impl TomlBuildSystem {
    pub fn from_toml_str(source: &str) -> Result<Self, TomlError> {
        toml_edit::de::from_str(source).map_err(TomlError::from)
    }

    pub fn into_build_system(self) -> Result<BuildSystem, TomlError> {
        // Parse the build backend and ensure it is a binary spec.
        let build_backend_spec = self
            .build_backend
            .value
            .spec
            .into_binary_spec()
            .map_err(|e| {
                TomlError::Generic(e.to_string().into(), self.build_backend.span.clone())
            })?;

        // Convert the additional dependencies and make sure that they are binary.
        let additional_dependencies = self
            .additional_dependencies
            .specs
            .into_iter()
            .map(|(name, spec)| match spec.into_source_or_binary() {
                Either::Right(binary) => Ok((name, binary)),
                Either::Left(_source) => {
                    let spec_range = self
                        .additional_dependencies
                        .value_spans
                        .get(&name)
                        .or_else(|| self.additional_dependencies.name_spans.get(&name))
                        .cloned();
                    Err(TomlError::Generic(
                        "Cannot use source dependencies for build backends dependencies".into(),
                        spec_range,
                    ))
                }
            })
            .collect::<Result<_, TomlError>>()?;

        // Make sure there are no empty channels
        if let Some(channels) = &self.channels {
            if channels.value.is_empty() {
                return Err(TomlError::Generic(
                    "No channels specified for the build backend dependencies".into(),
                    channels.span(),
                ));
            }
        }

        Ok(BuildSystem {
            build_backend: BuildBackend {
                name: self.build_backend.value.name.value,
                spec: build_backend_spec,
                additional_args: None,
            },
            additional_dependencies,
            channels: self.channels.map(|channels| channels.value),
        })
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;

    use super::*;
    use crate::utils::test_utils::format_parse_error;

    fn expect_parse_failure(pixi_toml: &str) -> String {
        let parse_error = TomlBuildSystem::from_toml_str(pixi_toml)
            .and_then(TomlBuildSystem::into_build_system)
            .expect_err("parsing should fail");

        format_parse_error(pixi_toml, parse_error)
    }

    #[test]
    fn test_disallow_source() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", git = "https://github.com/org/repo" }
        "#
        ));
    }

    #[test]
    fn test_missing_version_specifier() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar" }
        "#
        ));
    }

    #[test]
    fn test_missing_backend() {
        assert_snapshot!(expect_parse_failure(""));
    }

    #[test]
    fn test_missing_name() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { version = "12.*" }
        "#
        ));
    }

    #[test]
    fn test_empty_channels() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*" }
            channels = []
        "#
        ));
    }

    #[test]
    fn test_additional_build_backend_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*", foo = "bar" }
        "#
        ));
    }

    #[test]
    fn test_additional_keys() {
        assert_snapshot!(expect_parse_failure(
            r#"
            build-backend = { name = "foobar", version = "*" }
            additional = "key"
        "#
        ));
    }
}
