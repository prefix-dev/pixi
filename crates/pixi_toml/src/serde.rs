//! Provides a bridge to deserialize types that implement `serde::Deserialize` but not `toml_span::Deserialize`.

use std::collections::BTreeMap;

use serde::Deserialize;
use toml_span::{DeserError, ErrorKind, Value, value::ValueInner};

use crate::DeserializeAs;

/// A wrapper type that enables deserializing any type that implements `serde::Deserialize`
/// from a TOML value.
///
/// This is useful when you have types from external crates that already implement
/// `serde::Deserialize` but not `toml_span::Deserialize`.
///
/// # Example
/// ```ignore
/// use pixi_toml::{Serde, TomlWith};
///
/// // Assuming ExternalType implements serde::Deserialize but not toml_span::Deserialize
/// let value = th.optional::<TomlWith<_, Serde<ExternalType>>>("field")
///     .map(TomlWith::into_inner);
/// ```
pub struct Serde<T>(std::marker::PhantomData<T>);

impl<'de, T> DeserializeAs<'de, T> for Serde<T>
where
    T: Deserialize<'de>,
{
    fn deserialize_as(value: &mut Value<'de>) -> Result<T, DeserError> {
        // Convert TOML value to serde value
        let value_as_json = convert_toml_to_serde(value)?;

        // Deserialize using T's serde::Deserialize implementation
        T::deserialize(serde_value::ValueDeserializer::new(value_as_json)).map_err(
            |e: serde_value::DeserializerError| {
                DeserError::from(toml_span::Error {
                    kind: ErrorKind::Custom(e.to_string().into()),
                    span: value.span,
                    line_info: None,
                })
            },
        )
    }
}

/// Convert a TOML value to a serde value
pub fn convert_toml_to_serde(value: &mut Value) -> Result<serde_value::Value, DeserError> {
    Ok(match value.take() {
        ValueInner::String(s) => serde_value::Value::String(s.to_string()),
        ValueInner::Integer(i) => serde_value::Value::I64(i),
        ValueInner::Float(f) => serde_value::Value::F64(f),
        ValueInner::Boolean(b) => serde_value::Value::Bool(b),
        ValueInner::Array(mut arr) => {
            let mut json_arr = Vec::new();
            for item in &mut arr {
                json_arr.push(convert_toml_to_serde(item)?);
            }
            serde_value::Value::Seq(json_arr)
        }
        ValueInner::Table(table) => {
            let mut map = BTreeMap::new();
            for (key, mut val) in table {
                map.insert(
                    serde_value::Value::String(key.to_string()),
                    convert_toml_to_serde(&mut val)?,
                );
            }
            serde_value::Value::Map(map)
        }
    })
}
