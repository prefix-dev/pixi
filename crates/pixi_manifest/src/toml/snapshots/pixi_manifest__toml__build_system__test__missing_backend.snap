---
source: crates/pixi_manifest/src/toml/build_system.rs
expression: "expect_parse_failure(\"\")"
---
  × missing field `build-backend`
   ╭─[pixi.toml:1:1]
   ╰────
